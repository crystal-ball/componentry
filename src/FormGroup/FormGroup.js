import { staticComponent } from '../utils/static-component-builder'

/**
 * [FormGroup component 📝](https://componentry.design/components/form-group)
 */
export const FormGroup = staticComponent('FormGroup', {
  componentCx: 'form-group',
})
