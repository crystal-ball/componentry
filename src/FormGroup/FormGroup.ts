/* eslint-disable @typescript-eslint/no-empty-interface */

import { ComponentBaseProps } from '../utils/types'
import { staticComponent } from '../utils/static-component-builder'

export interface FormGroupProps extends ComponentBaseProps<'div'> {}

/**
 * [FormGroup component üìù](https://componentry.design/components/form-group)
 * @experimental
 */
export const FormGroup = staticComponent('FormGroup')
