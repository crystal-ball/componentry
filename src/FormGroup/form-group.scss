/* componentry/src/FormGroup/form-group */

.ðŸ…²-form-group {
  margin-bottom: $form-group-margin-bottom;
}
