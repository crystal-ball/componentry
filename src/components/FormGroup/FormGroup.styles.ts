import { getMergedConfig } from '../../plugin-postcss/configs'

const { theme } = getMergedConfig()

//                                     <FormGroup /> styles
// --------------------------------------------------------

export const FormGroup = {
  '.ðŸ…²FormGroup': {
    marginBottom: theme.spacing[2],
  },
}
