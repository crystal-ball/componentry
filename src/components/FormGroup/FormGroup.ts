import { type ComponentBaseProps } from '../../utils/base-types'
import { staticComponent } from '../../utils/static-component-builder'

export interface FormGroupProps extends ComponentBaseProps<'div'> {}

/**
 * [FormGroup component üìù](https://componentry.design/components/form-group)
 * @experimental
 */
export const FormGroup = staticComponent('FormGroup')
