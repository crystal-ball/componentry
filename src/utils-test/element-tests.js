/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
