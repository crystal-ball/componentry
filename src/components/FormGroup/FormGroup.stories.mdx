import { Meta, Story } from '@storybook/addon-docs'

import { Input } from '../Input/Input'
import { FormGroup } from './FormGroup'

<Meta title='Components/FormGroup' component={FormGroup} />

# FormGroup

Form groups provide a control point for standardizing spacing within forms.

<Story name='FormGroup'>
  <div className='w-1/2'>
    <FormGroup>
      <Input>
        <Input.Label>Name</Input.Label>
        <Input.Field placeholder='' />
      </Input>
    </FormGroup>
    <FormGroup>
      <Input>
        <Input.Label>Email</Input.Label>
        <Input.Field placeholder='' />
      </Input>
    </FormGroup>
  </div>
</Story>
