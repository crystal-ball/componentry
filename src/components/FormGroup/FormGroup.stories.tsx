import type { Meta, StoryObj } from '@storybook/react'

import { Input } from '../Input/Input'
import { FormGroup } from './FormGroup'

const meta: Meta<typeof FormGroup> = {
  component: FormGroup,
}

export default meta
type Story = StoryObj<typeof FormGroup>

export const Primary: Story = {
  render: () => (
    <div className='w-1/2'>
      {/* @ts-expect-error -- experimental component */}
      <FormGroup>
        <Input>
          <Input.Label>Name</Input.Label>
          <Input.Field placeholder='' />
        </Input>
      </FormGroup>
      {/* @ts-expect-error -- experimental component */}
      <FormGroup>
        <Input>
          <Input.Label>Email</Input.Label>
          <Input.Field placeholder='' />
        </Input>
      </FormGroup>
    </div>
  ),
}
