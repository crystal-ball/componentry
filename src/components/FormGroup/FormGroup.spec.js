import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { FormGroup } from './FormGroup'

describe('<FormGroup />', () => {
  // Basic library element test suite
  elementTests(FormGroup)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<FormGroup /> snapshots', () => {
  it('renders correctly', () => {
    render(<FormGroup data-testid='formgroup' />)

    expect(screen.getByTestId('formgroup')).toMatchSnapshot()
  })
})
