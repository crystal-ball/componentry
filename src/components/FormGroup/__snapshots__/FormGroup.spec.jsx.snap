// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<FormGroup /> snapshots > renders correctly 1`] = `
<div
  class="C9Y-FormGroup"
  data-testid="formgroup"
/>
`;
